sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala

```diff
@@ -0,0 +1,577 @@
+          true
+        case char@_ if char >= ZERO_DIGIT && char <= NINE_DIGIT =>
+          numDigits += 1
+          // Append each group of input digits to the appropriate before/afterDecimalPoint
+          // string for later use in constructing the result Decimal value.
+          if (reachedDecimalPoint) {
+            afterDecimalPoint.append(char)
+          } else {
+            beforeDecimalPoint.append(char)
+          }
+          true
+        case COMMA_SIGN =>
+          actualDigitLengths.prepend(numDigits)
+          numDigits = 0
+          true
+        case _ =>
+          actualDigitLengths.prepend(numDigits)
+          false
+      })) {
+      inputIndex += 1
+    }
+    if (inputIndex == inputLength) {
+      actualDigitLengths.prepend(numDigits)
+    }
+    // Compare the number of digits encountered in each group (separated by thousands
+    // separators) with the expected numbers from the format string.
+    if (actualDigitLengths.length > expectedDigits.length) {
+      // The input contains more thousands separators than the format string.
+      return None
+    }
+    for (i <- 0 until expectedDigits.length) {
+      val expectedToken: Digits = expectedDigits(i)
+      val actualNumDigits: Int =
+        if (i < actualDigitLengths.length) {
+          actualDigitLengths(i)
+        } else {
+          0
+        }
+      expectedToken match {
+        case ExactlyAsManyDigits(expectedNumDigits)
+          if actualNumDigits != expectedNumDigits =>
+          // The input contained more or fewer digits than required.
+          return None
+        case AtMostAsManyDigits(expectedMaxDigits)
+          if actualNumDigits > expectedMaxDigits =>
+          // The input contained more digits than allowed.
+          return None
+        case _ =>
+      }
+    }
+    Some(inputIndex)
+  }
+
+  /**
+   * This method executes when the input string fails to match the format string. It throws an
+   * exception if indicated on construction of this class, or returns NULL otherwise.
+   */
+  private def formatMatchFailure(input: UTF8String, originNumberFormat: String): Decimal = {
+    if (errorOnFail) {
+      throw QueryExecutionErrors.invalidNumberFormatError(input, originNumberFormat)
+    }
+    null
+  }
+
+  /**
+   * Computes the final Decimal value from the beforeDecimalPoint and afterDecimalPoint fields of
+   * this class, as a result of parsing.
+   *
+   * @param negateResult whether the input string specified to negate the result
+   * @return a Decimal value with the value indicated by the input string and the precision and
+   *         scale indicated by the format string
+   */
+  private def getDecimal(negateResult: Boolean): Decimal = {
+    // Consume all digits before the decimal point into the unscaled value.
+    var unscaled: Long = 0
+    if (beforeDecimalPoint.length > 0) {
+      unscaled += beforeDecimalPoint.toString.toLong
```

Comment:
shall we check if the string can fit a long type?

Reply:
I think a simple way is to build the number string and create Decimal
```
val numStr = ...
val javaDecimal = new java.math.BigDecimal(numStr)
if (precision <= Decimal.MAX_LONG_DIGITS) {
  // Constructs a `Decimal` with an unscaled `Long` value if possible.
  Decimal(javaDecimal.unscaledValue().longValue(), precision, scale)
} else {
  // Otherwise, resorts to an unscaled `BigInteger` instead.
  Decimal(javaDecimal, precision, scale)
}
```